exercises/grep/grep_test.rb

```diff
@@ -0,0 +1,402 @@
+      In such a presence here to plead my thoughts;
+      But I beseech your grace that I may know
+      The worst that may befall me in this case,
+      If I refuse to wed Demetrius.
+    END
+
+    IO.write 'paradise-lost.txt', <<~END
+      Of Mans First Disobedience, and the Fruit
+      Of that Forbidden Tree, whose mortal tast
+      Brought Death into the World, and all our woe,
+      With loss of Eden, till one greater Man
+      Restore us, and regain the blissful Seat,
+      Sing Heav'nly Muse, that on the secret top
+      Of Oreb, or of Sinai, didst inspire
+      That Shepherd, who first taught the chosen Seed
+    END
+  end
+
+  def teardown
+    File.delete('iliad.txt')
+    File.delete('midsummer-night.txt')
+    File.delete('paradise-lost.txt')
+  end
+
+  def test_one_file_one_match_no_flags
+    # skip
+    pattern = "Agamemnon"
+    flags = []
+    files = ["iliad.txt"]
+    expected = <<~EXPECTED.rstrip
+      Of Atreus, Agamemnon, King of men.
+    EXPECTED
+
+    assert_equal expected, Grep.grep(pattern, flags, files)
+  end
+
+  def test_one_file_one_match_print_line_numbers_flag
+    skip
+    pattern = "Forbidden"
+    flags = ["-n"]
+    files = ["paradise-lost.txt"]
+    expected = <<~EXPECTED.rstrip
+      2:Of that Forbidden Tree, whose mortal tast
+    EXPECTED
+
+    assert_equal expected, Grep.grep(pattern, flags, files)
+  end
+
+  def test_one_file_one_match_case_insensitive_flag
+    skip
+    pattern = "FORBIDDEN"
+    flags = ["-i"]
+    files = ["paradise-lost.txt"]
+    expected = <<~EXPECTED.rstrip
+      Of that Forbidden Tree, whose mortal tast
+    EXPECTED
+
+    assert_equal expected, Grep.grep(pattern, flags, files)
+  end
+
+  def test_one_file_one_match_print_file_names_flag
+    skip
+    pattern = "Forbidden"
+    flags = ["-l"]
+    files = ["paradise-lost.txt"]
+    expected = <<~EXPECTED.rstrip
+      paradise-lost.txt
+    EXPECTED
+
+    assert_equal expected, Grep.grep(pattern, flags, files)
+  end
+
+  def test_one_file_one_match_match_entire_lines_flag
+    skip
+    pattern = "With loss of Eden, till one greater Man"
+    flags = ["-x"]
+    files = ["paradise-lost.txt"]
+    expected = <<~EXPECTED.rstrip
+      With loss of Eden, till one greater Man
+    EXPECTED
+
+    assert_equal expected, Grep.grep(pattern, flags, files)
+  end
+
+  def test_one_file_one_match_multiple_flags
+    skip
+    pattern = "OF ATREUS, Agamemnon, KIng of MEN."
+    flags = ["-n", "-i", "-x"]
+    files = ["iliad.txt"]
+    expected = <<~EXPECTED.rstrip
+      9:Of Atreus, Agamemnon, King of men.
+    EXPECTED
+
+    assert_equal expected, Grep.grep(pattern, flags, files)
+  end
+
+  def test_one_file_several_matches_no_flags
+    skip
+    pattern = "may"
+    flags = []
+    files = ["midsummer-night.txt"]
+    expected = <<~EXPECTED.rstrip
+      Nor how it may concern my modesty,
+      But I beseech your grace that I may know
+      The worst that may befall me in this case,
+    EXPECTED
+
```

Comment:
Thinking TDD, would it make sense to put all the no-flags tests at the beginning?

Reply:
The order is defined by the canonical data. Do we want to go against that? (It may be worth considering opening a PR on the problem-specifications repo if you feel strongly about changing the order everywhere).

Reply:
Aha, no, we don't want to go against that. And as the TDD sequence may differ from language to language, I don't think it's worth to change it in the prob-specs.